daprovider/referenceda/reference_reader.go

```diff
@@ -0,0 +1,154 @@
+// Copyright 2021-2025, Offchain Labs, Inc.
+// For license information, see https://github.com/OffchainLabs/nitro/blob/master/LICENSE.md
+
+package referenceda
+
+import (
+	"context"
+	"crypto/sha256"
+	"fmt"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/log"
+
+	"github.com/offchainlabs/nitro/arbutil"
+	"github.com/offchainlabs/nitro/daprovider"
+	"github.com/offchainlabs/nitro/util/containers"
+)
+
+// Reader implements the daprovider.Reader interface for ReferenceDA
+type Reader struct {
+	storage       *InMemoryStorage
+	l1Client      *ethclient.Client
+	validatorAddr common.Address
+}
+
+// NewReader creates a new ReferenceDA reader
+func NewReader(storage *InMemoryStorage, l1Client *ethclient.Client, validatorAddr common.Address) *Reader {
+	return &Reader{
+		storage:       storage,
+		l1Client:      l1Client,
+		validatorAddr: validatorAddr,
+	}
+}
+
+// recoverInternal is the shared implementation for both RecoverPayload and CollectPreimages
+func (r *Reader) recoverInternal(
+	ctx context.Context,
+	batchNum uint64,
+	batchBlockHash common.Hash,
+	sequencerMsg []byte,
+	needPayload bool,
+	needPreimages bool,
+) ([]byte, daprovider.PreimagesMap, error) {
+	if len(sequencerMsg) <= 40 {
+		return nil, nil, fmt.Errorf("sequencer message too small")
+	}
+
+	// Skip the 40-byte L1 header and get the certificate
+	certBytes := sequencerMsg[40:]
+
+	// Deserialize certificate
+	cert, err := Deserialize(certBytes)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed to deserialize certificate: %w", err)
+	}
+
+	// Validate certificate - always validate for ReferenceDA
+	// TODO: Uncomment the following once we have merged customda contracts changes.
```

Comment:
can we add the ReferenceDA contract in contract-local, which is used for testing?
I know we need changes in the contracts as well, but the referenceDA should probably only be here.